src/Tutor.Core/InstructorModel/Instructors/DefaultInstructor.cs

```diff
@@ -1,17 +1,93 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
 using FluentResults;
 using Tutor.Core.DomainModel.AssessmentEvents;
+using Tutor.Core.DomainModel.KnowledgeComponents;
 
 namespace Tutor.Core.InstructorModel.Instructors
 {
     public class DefaultInstructor : IInstructor
     {
-        public DefaultInstructor()
+        private readonly IKCRepository _ikcRepository;
+        private readonly IAssessmentEventRepository _assessmentEventRepository;
+
+        public DefaultInstructor(IKCRepository ikcRepository, IAssessmentEventRepository assessmentEventRepository)
         {
+            _ikcRepository = ikcRepository;
+            _assessmentEventRepository = assessmentEventRepository;
         }
 
         public Result<AssessmentEvent> SelectSuitableAssessmentEvent(int knowledgeComponentId, int learnerId)
         {
-            throw new System.NotImplementedException();
+            if (_ikcRepository.GetAssessmentEventsByKnowledgeComponent(knowledgeComponentId).Count == 0) return Result.Fail("There are no assessment events yet.");
+            var suitableAssessmentEvent = GetAssessmentEventWithoutSubmission(knowledgeComponentId, learnerId);
+            return Result.Ok(suitableAssessmentEvent != null
+                ? _assessmentEventRepository.GetAssessmentEvent(suitableAssessmentEvent.Id)
+                : _assessmentEventRepository.GetAssessmentEvent(
+                    FindSubmissionWithMinCorrectness(knowledgeComponentId, learnerId).AssessmentEventId));
+        }
+
+        public void UpdateKcMastery(Submission submission, int knowledgeComponentId)
+        {
+            var currentCorrectnessLevel = _assessmentEventRepository
+                .FindSubmissionWithMaxCorrectness(submission)?.CorrectnessLevel ?? 0.0;
+            if (currentCorrectnessLevel > submission.CorrectnessLevel) return;
+
+            var kcMastery = _ikcRepository.GetKnowledgeComponentMastery(submission.LearnerId, knowledgeComponentId);
+
+            var kcMasteryIncrement = 100.0 / _ikcRepository.GetAssessmentEventsByKnowledgeComponent(knowledgeComponentId).Count
```

Comment:
Instead of loading all AEs to count them, you can create a repository function that simply sends a count SQL query (e.g., return _dbContext.AEs.Count(ae => ae.KCid == kcId). Should be faster.